cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -0,0 +1,572 @@
+using internal::Bitmap;
+
+namespace compute {
+
+namespace {
+
+util::optional<uint64_t> GetConstantValidityWord(const Datum& data) {
+  if (data.is_scalar()) {
+    return data.scalar()->is_valid ? ~uint64_t(0) : uint64_t(0);
+  }
+
+  if (data.array()->null_count == data.array()->length) return uint64_t(0);
+
+  if (!data.array()->MayHaveNulls()) return ~uint64_t(0);
+
+  // no constant validity word available
+  return {};
+}
+
+inline Bitmap GetBitmap(const Datum& datum, int i) {
+  if (datum.is_scalar()) return {};
+  const ArrayData& a = *datum.array();
+  return Bitmap{a.buffers[i], a.offset, a.length};
+}
+
+// if the condition is null then output is null otherwise we take validity from the
+// selected argument
+// ie. cond.valid & (cond.data & left.valid | ~cond.data & right.valid)
+Status PromoteNullsVisitor(KernelContext* ctx, const Datum& cond_d, const Datum& left_d,
+                           const Datum& right_d, ArrayData* output) {
+  auto cond_const = GetConstantValidityWord(cond_d);
+  auto left_const = GetConstantValidityWord(left_d);
+  auto right_const = GetConstantValidityWord(right_d);
+
+  enum { COND_CONST = 1, LEFT_CONST = 2, RIGHT_CONST = 4 };
+  auto flag = COND_CONST * cond_const.has_value() | LEFT_CONST * left_const.has_value() |
+              RIGHT_CONST * right_const.has_value();
+
+  const ArrayData& cond = *cond_d.array();
+  // cond.data will always be available
+  Bitmap cond_data{cond.buffers[1], cond.offset, cond.length};
+  Bitmap cond_valid{cond.buffers[0], cond.offset, cond.length};
+  Bitmap left_valid = GetBitmap(left_d, 0);
+  Bitmap right_valid = GetBitmap(right_d, 0);
+  // sometimes Bitmaps will be ignored, in which case we replace access to them with
+  // duplicated (probably elided) access to cond_data
+  const Bitmap& _ = cond_data;
+
+  // lambda function that will be used inside the visitor
+  uint64_t* out_validity = nullptr;
+  int64_t i = 0;
+  auto apply = [&](uint64_t c_valid, uint64_t c_data, uint64_t l_valid,
+                   uint64_t r_valid) {
+    out_validity[i] = c_valid & ((c_data & l_valid) | (~c_data & r_valid));
+    i++;
+  };
+
+  // if the condition is null then output is null otherwise we take validity from the
+  // selected argument
+  // ie. cond.valid & (cond.data & left.valid | ~cond.data & right.valid)
+
+  // In the following cases, we dont need to allocate out_valid bitmap
+  switch (flag) {
+    case COND_CONST | LEFT_CONST | RIGHT_CONST:
+      // if cond & left & right all ones, then output is all valid --> out_valid = nullptr
+      if ((*cond_const & *left_const & *right_const) == UINT64_MAX) {
+        return Status::OK();
+      }
+    case LEFT_CONST | RIGHT_CONST:
+      // if both left and right are valid, no need to calculate out_valid bitmap. Pass
+      // cond validity buffer
+      if ((*left_const & *right_const) == UINT64_MAX) {
+        output->buffers[0] = SliceBuffer(cond.buffers[0], cond.offset, cond.length);
```

Comment:
I don't think this does what you want: cond.offset/cond.length are bitwise here but SliceBuffer is bytewise. You probably do need to allocate a validity bitmap and copy it using CopyBitmap.

Reply:
I see! Thanks David!

Reply:
I don't test arrays with offsets. I think it would be better to add some test cases there